pallets/frequency-tx-payment/src/lib.rs

```diff
@@ -1,34 +1,96 @@
 #![cfg_attr(not(feature = "std"), no_std)]
 
 use frame_support::{
-	dispatch::{DispatchInfo, GetDispatchInfo, PostDispatchInfo},
+	dispatch::{DispatchInfo, Dispatchable, GetDispatchInfo, PostDispatchInfo},
 	pallet_prelude::*,
-	traits::IsType,
+	traits::{IsSubType, IsType},
+	DefaultNoBound,
 };
+use frame_system::pallet_prelude::*;
 
 use codec::{Decode, Encode};
 use scale_info::TypeInfo;
 use sp_std::prelude::*;
 
 use sp_runtime::{
-	traits::{DispatchInfoOf, Dispatchable, PostDispatchInfoOf, SignedExtension},
+	traits::{DispatchInfoOf, PostDispatchInfoOf, SignedExtension, Zero},
 	transaction_validity::{TransactionValidity, TransactionValidityError},
 	FixedPointOperand,
 };
 
+use common_primitives::capacity::Nontransferable;
+
 use pallet_transaction_payment::OnChargeTransaction;
 
-// Type aliases used for interaction with `OnChargeTransaction`.
+/// Type aliases used for interaction with `OnChargeTransaction`.
 pub(crate) type OnChargeTransactionOf<T> =
 	<T as pallet_transaction_payment::Config>::OnChargeTransaction;
 
-// Balance type alias.
+/// Balance type alias.
 pub(crate) type BalanceOf<T> = <OnChargeTransactionOf<T> as OnChargeTransaction<T>>::Balance;
 
-// Liquidity info type alias (imbalances).
+/// Liquidity info type alias (imbalances).
 pub(crate) type LiquidityInfoOf<T> =
 	<OnChargeTransactionOf<T> as OnChargeTransaction<T>>::LiquidityInfo;
 
+/// Capacity Balance type
+pub(crate) type CapacityOf<T> = <T as Config>::Capacity;
+
+/// Capacity Balance alias
+pub(crate) type CapacityBalanceOf<T> = <CapacityOf<T> as Nontransferable>::Balance;
+
+/// Used to pass the initial payment info from pre- to post-dispatch.
+#[derive(Encode, Decode, DefaultNoBound, TypeInfo)]
+pub enum InitialPayment<T: Config> {
+	/// No initial fee was payed.
+	#[default]
+	Free,
+	/// The initial fee was payed in the native currency.
+	Token(LiquidityInfoOf<T>),
+	/// The initial fee was payed in an asset.
+	Capacity,
+}
+
+#[cfg(feature = "std")]
+impl<T: Config> InitialPayment<T> {
+	pub fn is_free(&self) -> bool {
+		match *self {
+			InitialPayment::Free => true,
+			_ => false,
+		}
```

Comment:
I like match as well, but not sure why not to just derive Eq and use `==` for the smaller result
```suggestion
		self == InitialPayment::Free
```

Reply:
I did this because I could not add the `==.` It was a workaround since `LiquidityInfoOf<T>` does not implement
equals. LiquidityInfoOf<T> is an alias for something from the original transaction-payment pallet.

These are just for testing and are used in std only.

```
pub enum InitialPayment<T: Config> {
	/// No initial fee was paid.
	#[default]
	Free,
	/// The initial fee was paid in the native currency.
	Token(LiquidityInfoOf<T>),
	/// The initial fee was paid in an asset.
	Capacity,
}
```